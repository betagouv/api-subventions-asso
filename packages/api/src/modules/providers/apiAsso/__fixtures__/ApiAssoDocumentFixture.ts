import { StructureDacDocumentDto, StructureRnaDocumentDto } from "../dto/StructureDto";

export const ApiAssoDocumentFixture = {
    asso: {
        documents: {
            nbDocDac: 1,
            document_dac: [
                {
                    url: "/fake/url",
                    nom: "nom fake",
                    meta: {
                        type: "BPA",
                        id_siret: 50922194100000,
                        etat: "courant",
                    },
                    time_depot: "2021-06-18 12:02:53",
                },
            ] as StructureDacDocumentDto[],
            nbDocRna: 1,
            document_rna: [
                {
                    url: "/fake/url",
                    type: "PV",
                    sous_type: "PV",
                    id: "ididididid",
                    lib_sous_type: "PV",
                    time: 1622557171,
                    annee: 2021,
                },
            ] as StructureRnaDocumentDto[],
        },
    },
};
