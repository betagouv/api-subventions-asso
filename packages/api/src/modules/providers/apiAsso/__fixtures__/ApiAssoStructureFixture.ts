import StructureDto, 
{ 
    StructureDacDocumentDto,
    StructureEtablissementDto,
    StructureRepresentantLegalDto,
    StructureRibDto,
    StructureRnaDocumentDto
} from "../dto/StructureDto"

export const fixtureEtablissements = [
    {
        id_siret: "50922194100000",
        adresse: {},
    },
    {
        id_siret: "50922194100001",
        adresse: {},
    }
] as StructureEtablissementDto[]

export const fixtureDocumentRna = [{
    url: "/fake/url",
    type: "PV",
    'sous-type': "PV",
    id: "ididididid",
    time: 1622557171,
    annee: 2021,
}] as StructureRnaDocumentDto[]

export const fixtureDocumentDac = [{
    url: "/fake/url",
    nom: "nom fake",
    meta: {
        type: "BPA",
        id_siret: "50922194100000",
        etat: "courant"
    },
    time_depot: "2021-06-18 12:02:53"
}] as StructureDacDocumentDto[]

export const fixtureRib = [
    {
        id_siret: "50922194100000",
        iban: "TEST",
        bic: "BIC_TEST"
    },
    {
        id_siret: "50922194100000",
        iban: "TEST 2",
        bic: "BIC_TEST 2",
        url: "fake/path/to/rib"
    }
] as StructureRibDto[];

export const fixtureRepresentantLegal = [{
    nom: "Jedusor",
    prenom: "Tom",
    role: "Mage pas simpa",
    id_siret: "50922194100000"
}] as unknown as StructureRepresentantLegalDto[]

export const fixtureAsso = {
    identite: {
        nom: "TEST",
        nom_sirene: "TEST SIREN 2",
        id_rna: "W00000000",
        id_siren: "509221941",
        id_siret_siege: "5092219410000",
        id_forme_juridique: "9220",
        date_modif_rna: "01-01-2022",
        date_modif_siren: "01-01-2022",
        regime: "Autre"
    },
    activites: {},
    coordonnees: {
        adresse_siege: {
            commune: "paris"
        }
    },
    rib: fixtureRib,
    etablissement: fixtureEtablissements,
    document_rna: fixtureDocumentRna,
    document_dac: fixtureDocumentDac
} as unknown as StructureDto