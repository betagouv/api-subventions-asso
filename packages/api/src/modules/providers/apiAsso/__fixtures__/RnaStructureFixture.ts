import { RnaStructureDto } from "../dto/RnaStructureDto";

export const rnaStructureFixture: RnaStructureDto = {
    identite: {
        nom: "Asso qui n'existe pas",
        sigle: "Asso qui n'existe p...",
        id_rna: "W000000000",
        date_pub_jo: "1998-02-07",
        date_modif_rna: "2023-01-12",
        active: true,
        nature: "Simplement déclarée"
    },
    activites: {
        objet: "Object social !",
        lib_objet_social1: "conduite d'activités économiques",
        annee_activite_principale: 0,
        effectif_salarie_cent: 0,
        annee_effectif_salarie_cent: 0
    },
    coordonnees: {
        adresse_siege: {
            cplt_1: "Place des orties",
            num_voie: "18",
            voie: "MEURIS",
            commune: "Londre",
            cp: "42424",
            type_voie: "RUE"
        },
        adresse_gestion: {
            cplt_1: "Place des orties",
            voie: "MEURIS",
            commune: "Londre",
            cp: "42424",
            pays: "Groenland"
        }
    }
};
