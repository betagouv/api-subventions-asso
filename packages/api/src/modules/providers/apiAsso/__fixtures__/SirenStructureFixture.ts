import { SirenStructureDto } from "../dto/SirenStructureDto";

export const sirenStructureFixture: SirenStructureDto = {
    identite: {
        nom: "Je suis un fause asso",
        nom_sirene: "Je suis un fause asso",
        id_siren: 1230046480,
        id_siret_siege: 123004648000021,
        id_forme_juridique: 9220,
        date_creation_sirene: "1993-01-11",
        date_modif_siren: "2019-11-28",
        active: true,
        active_sirene: true,
        regime: "loi1901",
    },
    activites: {
        id_activite_principale: "94.99Z",
        annee_activite_principale: 0,
        id_tranche_effectif: "NN",
        effectif_salarie_cent: 0,
        annee_effectif_salarie_cent: 0,
        appartenance_ess: "O",
    },
    coordonnees: {
        adresse_siege: {
            cplt_1: "Place des orties",
            num_voie: 18,
            voie: "MEURIS",
            commune: "Londre",
            code_insee: 42424,
            cp: 42424,
            type_voie: "RUE",
        },
        adresse_gestion: {
            cplt_1: "Place des orties",
            num_voie: 18,
            voie: "MEURIS",
            commune: "Londre",
            code_insee: 42424,
            cp: 42424,
            type_voie: "RUE",
        },
    },
    nbEtabsActifs: 1,
    etablissement: [
        {
            actif: true,
            adresse: {
                cplt_1: "Place des orties",
                num_voie: 18,
                voie: "MEURIS",
                commune: "Londre",
                code_insee: 42424,
                cp: 42424,
                type_voie: "RUE",
            },
            id_siret: 123004648000021,
            est_siege: true,
            date_actif: "2019-01-02",
            id_activite_principale: "94.99Z",
            annee_activite_principale: 0,
            effectif_salarie_cent: 0,
            annee_effectif_salarie_cent: 0,
        },
        {
            actif: false,
            adresse: {
                num_voie: 221,
                voie: "Baker",
                commune: "London",
                code_insee: 77777,
                cp: 77777,
                type_voie: "St",
            },
            id_siret: 123004648000013,
            est_siege: false,
            date_actif: "1993-01-10",
            id_activite_principale: "94.99Z",
            annee_activite_principale: 0,
            id_tranche_effectif: "NN",
            effectif_salarie_cent: 0,
            annee_effectif_salarie_cent: 0,
        },
    ],
};
