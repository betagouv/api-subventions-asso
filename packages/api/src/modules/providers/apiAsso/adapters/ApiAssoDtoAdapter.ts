import { siretToNIC } from "../../../../shared/helpers/SirenHelper";
import ProviderValueFactory from "../../../../shared/ProviderValueFactory";
import { Etablissement, Association }from "@api-subventions-asso/dto";
import Document from "../../../documents/@types/Document";
import StructureDto, { StructureDacDocumentDto, StructureEtablissementDto, StructureRepresentantLegalDto, StructureRibDto, StructureRnaDocumentDto } from "../dto/StructureDto";

export default class ApiAssoDtoAdapter {
    static providerNameRna = "BASE RNA <Via API ASSO>";
    static providerNameSiren = "BASE SIREN <Via API ASSO>";

    static toAssociation(structure: StructureDto): Association[] {
        const associations: Association[] = []
        const fromRNA = structure.identite.regime === "loi1901"; // Data come from rna
        const toDate = (stringDate: string) => {
            const [year, month, day] = stringDate.split("-").map(string => parseInt(string, 10));
            return new Date(Date.UTC(year, month-1, day));
        }
        
        if (structure.identite.date_modif_rna) {
            const toRnaPvs = ProviderValueFactory.buildProviderValuesAdapter(this.providerNameRna, toDate(structure.identite.date_modif_rna));   
            const rnaAssociation: Association = {
                rna: toRnaPvs(structure.identite.id_rna),
                denomination: toRnaPvs(structure.identite.nom),
                date_creation: structure.identite.date_creat ? toRnaPvs(toDate(structure.identite.date_creat)) : undefined,
                date_modification: toRnaPvs(toDate(structure.identite.date_modif_rna)),
                objet_social: toRnaPvs(structure.activites.objet),
                code_objet_social_1: toRnaPvs(structure.activites.id_objet_social1),
                code_objet_social_2: toRnaPvs(structure.activites.id_objet_social2),
                adresse_siege : fromRNA ? toRnaPvs({
                    numero: structure.coordonnees.adresse_siege.num_voie,
                    type_voie: structure.coordonnees.adresse_siege.type_voie,
                    voie: structure.coordonnees.adresse_siege.voie,
                    code_postal: structure.coordonnees.adresse_siege.cp,
                    commune: structure.coordonnees.adresse_siege.commune,
                }) : undefined
            }
            associations.push(rnaAssociation);
        }
        
        if (structure.identite.date_modif_siren) {
            const toSirenPvs = ProviderValueFactory.buildProviderValuesAdapter(this.providerNameSiren, toDate(structure.identite.date_modif_siren));
            const adresse = structure.coordonnees.adresse_siege_sirene || fromRNA ? undefined : structure.coordonnees.adresse_siege;
            const sirenAssociation: Association = {
                denomination: structure.identite.nom_sirene ? toSirenPvs(structure.identite.nom_sirene): toSirenPvs(structure.identite.nom),
                siren: toSirenPvs(structure.identite.id_siren),
                nic_siege: toSirenPvs(siretToNIC(structure.identite.id_siret_siege)),
                categorie_juridique: toSirenPvs(structure.identite.id_forme_juridique.toString()),
                date_creation: structure.identite.date_creation_sirene ? toSirenPvs(toDate(structure.identite.date_creation_sirene)) : undefined,
                date_modification: toSirenPvs(toDate(structure.identite.date_modif_siren)),
                adresse_siege: adresse ? toSirenPvs({
                    numero: adresse.num_voie,
                    type_voie: adresse.type_voie,
                    voie: adresse.voie,
                    code_postal: adresse.cp,
                    commune: adresse.commune,
                }) : undefined,
                etablisements_siret: toSirenPvs(structure.etablissement.map(e => e.id_siret))
            }
            associations.push(sirenAssociation);
        }

        return associations
    }

    static toEtablissement(etablissement: StructureEtablissementDto, ribs: StructureRibDto[], representantsLegaux: StructureRepresentantLegalDto[], dateModif: string): Etablissement {
        const toDate = (stringDate: string) => {
            const [year, month, day] = stringDate.split("-").map(string => parseInt(string, 10));
            return new Date(Date.UTC(year, month-1, day));
        }
        const toSirenPvs = ProviderValueFactory.buildProviderValuesAdapter(this.providerNameSiren, toDate(dateModif));

        const toContact = (r: StructureRepresentantLegalDto) => (
            {
                nom: r.nom,
                prenom: r.prenom,
                civilite: r.civilitÃ©,
                telephone: r.telephone,
                email: r.courriel,
                role: r.fonction
            }
        )
        
        return {
            siret: toSirenPvs(etablissement.id_siret),
            nic: toSirenPvs(siretToNIC(etablissement.id_siret)),
            ouvert: toSirenPvs(etablissement.actif),
            siege: toSirenPvs(etablissement.est_siege),
            adresse: toSirenPvs({
                numero: etablissement.adresse.num_voie,
                type_voie: etablissement.adresse.type_voie,
                voie: etablissement.adresse.voie,
                code_postal: etablissement.adresse.cp,
                commune: etablissement.adresse.commune,
            }),
            information_banquaire: ribs.filter(rib => rib.id_siret === etablissement.id_siret).map(rib => toSirenPvs({ iban: rib.iban, bic: rib.bic})),
            representants_legaux: representantsLegaux ? representantsLegaux
                .filter(r => r.id_siret === etablissement.id_siret)
                .map(r => toSirenPvs(toContact(r))) : undefined,
            contacts: representantsLegaux ? representantsLegaux
                .filter(r => r.id_siret === etablissement.id_siret)
                .map(r => toSirenPvs(toContact(r))) : undefined,
        }
    }

    static toDocuments(structure: StructureDto): Document[] {
        const toDate = (stringDate: string) => {
            const [year, month, day] = stringDate.split("-").map(string => parseInt(string, 10));
            return new Date(Date.UTC(year, month-1, day));
        }

        const dataDate = toDate(structure.identite.date_modif_rna);
        const rnaDocuments = structure.document_rna?.map(document => ApiAssoDtoAdapter.rnaDocumentToDocument(document, dataDate)) || [];
        const dacDocuments = structure.document_dac?.map(document => ApiAssoDtoAdapter.dacDocumentToDocument(document, dataDate)) || [];
        const ribDocuments = structure.rib?.map(rib => ApiAssoDtoAdapter.ribDocumentToDocument(rib, dataDate)) || [];

        return [
            ...rnaDocuments,
            ...dacDocuments,
            ...ribDocuments.filter(r => r) as Document[]
        ]
    }

    private static rnaDocumentToDocument(rnaDocument: StructureRnaDocumentDto, date: Date): Document {
        const toRnaPv = ProviderValueFactory.buildProviderValueAdapter(this.providerNameRna, date);

        return {
            nom: toRnaPv(rnaDocument.id),
            type: toRnaPv(rnaDocument.type),
            url: toRnaPv(rnaDocument.url),
            __meta__: {}
        }
    }

    private static dacDocumentToDocument(dacDocument: StructureDacDocumentDto, date: Date): Document {
        const toRnaPv = ProviderValueFactory.buildProviderValueAdapter(this.providerNameRna, date);

        return {
            nom: toRnaPv(dacDocument.nom),
            type: toRnaPv(dacDocument.meta.type),
            url: toRnaPv(dacDocument.url),
            __meta__: {
                siret: dacDocument.meta.id_siret
            }
        }
    }

    private static ribDocumentToDocument(rib: StructureRibDto, date: Date): Document | null {
        if (!rib.url) return null;

        const toRnaPv = ProviderValueFactory.buildProviderValueAdapter(this.providerNameRna, date);

        return {
            nom: toRnaPv(rib.iban),
            type: toRnaPv("RIB"),
            url: toRnaPv(rib.url),
            __meta__: {
                siret: rib.id_siret
            }
        }
    }
}