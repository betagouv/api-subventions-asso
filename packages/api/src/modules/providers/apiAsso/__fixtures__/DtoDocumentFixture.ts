import ApiAssoDtoAdapter from "../adapters/ApiAssoDtoAdapter";

export const RnaDtoDocument = {
    __meta__: {},
    nom: {
        last_update: new Date("2021-01-19T18:42:37.171Z"),
        provider: ApiAssoDtoAdapter.providerNameRna,
        type: "string",
        value: "PV - ididididid"
    },
    type: {
        last_update: new Date("2021-01-19T18:42:37.171Z"),
        provider: ApiAssoDtoAdapter.providerNameRna,
        type: "string",
        value: "PV"
    },
    url: {
        last_update: new Date("2021-01-19T18:42:37.171Z"),
        provider: ApiAssoDtoAdapter.providerNameRna,
        type: "string",
        value: "/fake/url"
    }
};

export const DacSiret = "50922194100000";

export const DacDtoDocument = {
    __meta__: {
        siret: DacSiret
    },
    nom: {
        last_update: new Date("2021-06-18T12:02:00.000Z"),
        provider: ApiAssoDtoAdapter.providerNameLcaDocument,
        type: "string",
        value: "nom fake"
    },
    type: {
        last_update: new Date("2021-06-18T12:02:00.000Z"),
        provider: ApiAssoDtoAdapter.providerNameLcaDocument,
        type: "string",
        value: "BPA"
    },
    url: {
        last_update: new Date("2021-06-18T12:02:00.000Z"),
        provider: ApiAssoDtoAdapter.providerNameLcaDocument,
        type: "string",
        value: "/fake/url"
    }
};

export const DtoDocument = [RnaDtoDocument, DacDtoDocument];
