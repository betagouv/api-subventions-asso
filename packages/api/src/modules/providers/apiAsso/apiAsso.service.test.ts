import axios from "axios";
import { Association, Etablissement } from "@api-subventions-asso/dto"
import ApiAssoDtoAdapter from "./adapters/ApiAssoDtoAdapter";
import apiAssoService from "./apiAsso.service"
import { fixtureAsso, fixtureDocumentDac, fixtureDocumentRna, fixtureEtablissements } from "./__fixtures__/ApiAssoStructureFixture";

jest.mock('../../../shared/EventManager');

describe("ApiAssoService", () => {
    const axiosMock = jest.spyOn(axios, "get");
    const adapterAssoMock = jest.spyOn(ApiAssoDtoAdapter, "toAssociation").mockImplementation((r) => [{
        ...r,
        denomination_rna: [{ value: r.identite.nom, provider: "TEST" }],
        date_modification_rna: [{ value: new Date(Date.UTC(2022, 0, 1)) }]
    }] as unknown[] as Association[]);
    const adapterEtablissementMock = jest.spyOn(ApiAssoDtoAdapter, "toEtablissement").mockImplementation((r) => ({ ...r, siret: [{ value: r.id_siret }] }) as unknown as Etablissement);

    afterAll(() => {
        adapterAssoMock.mockReset();
        adapterEtablissementMock.mockReset();
    })

    describe("sendRequest", () => {
        // @ts-ignore
        const cache = apiAssoService.requestCache;
        const cacheHasMock = jest.spyOn(cache, "has");
        const cacheGetMock = jest.spyOn(cache, "get");

        afterAll(() => {
            cacheHasMock.mockClear();
            cacheGetMock.mockClear();
        })

        it("should return cache data", async () => {
            const expected = "FAKEDATA";

            cacheHasMock.mockImplementationOnce(() => true);
            cacheGetMock.mockImplementationOnce(() => [expected]);

            // @ts-ignore
            const actual = await apiAssoService.sendRequest("fake/route");
            expect(actual).toBe(expected);
        })

        it("should return api data", async () => {
            const expected = "FAKEDATA";
            axiosMock.mockImplementationOnce(() => Promise.resolve({
                status: 200,
                data: expected
            }))
            cacheHasMock.mockImplementationOnce(() => false);

            // @ts-ignore
            const actual = await apiAssoService.sendRequest("fake/route");
            expect(actual).toBe(expected);
        })


        it("should return null (wrong status code)", async () => {
            const expected = null;
            axiosMock.mockImplementationOnce(() => Promise.resolve({
                status: 404,
                data: 1
            }))
            cacheHasMock.mockImplementationOnce(() => false);

            // @ts-ignore
            const actual = await apiAssoService.sendRequest("fake/route");
            expect(actual).toBe(expected);
        })

        it("should return null (error is throw)", async () => {
            const expected = null;
            axiosMock.mockImplementationOnce(() => { throw new Error("Error test") });
            cacheHasMock.mockImplementationOnce(() => false);

            // @ts-ignore
            const actual = await apiAssoService.sendRequest("fake/route");
            expect(actual).toBe(expected);
        })
    })

    describe("findFullScopeAssociation", () => {
        // @ts-ignore
        const sirenCache = apiAssoService.dataSirenCache;
        const sirenCacheHasMock = jest.spyOn(sirenCache, "has");
        const sirenCacheGetMock = jest.spyOn(sirenCache, "get") as jest.SpyInstance<any>;
        const sirenCacheAddMock = jest.spyOn(sirenCache, "add") as jest.SpyInstance<any>;

        // @ts-ignore
        const rnaCache = apiAssoService.dataRnaCache;
        const rnaCacheHasMock = jest.spyOn(rnaCache, "has");
        const rnaCacheGetMock = jest.spyOn(rnaCache, "get") as jest.SpyInstance<any>;
        const rnaCacheAddMock = jest.spyOn(rnaCache, "add") as jest.SpyInstance<any>;

        // @ts-ignore
        const sendRequestMock = jest.spyOn(apiAssoService, "sendRequest") as jest.SpyInstance<any | null>;

        afterAll(() => {
            sirenCacheHasMock.mockClear();
            sirenCacheGetMock.mockClear();
            sirenCacheAddMock.mockClear();

            rnaCacheHasMock.mockClear();
            rnaCacheGetMock.mockClear();
            rnaCacheAddMock.mockClear();
        });

        it("should return data (cached by sirenCache)", async () => {
            const expected = "FAKE_DATA";
            sirenCacheHasMock.mockImplementationOnce(() => true);
            sirenCacheGetMock.mockImplementationOnce(() => [expected]);

            // @ts-ignore
            const actual = await apiAssoService.findFullScopeAssociation("ID");

            expect(actual).toBe(expected);
        })

        it("should return data (cached by rnaCache)", async () => {
            const expected = "FAKE_DATA";
            sirenCacheHasMock.mockImplementationOnce(() => false);
            rnaCacheHasMock.mockImplementationOnce(() => true);
            rnaCacheGetMock.mockImplementationOnce(() => [expected]);

            // @ts-ignore
            const actual = await apiAssoService.findFullScopeAssociation("ID");

            expect(actual).toBe(expected);
        })

        it("should return null", async () => {
            const expected = null;
            sirenCacheHasMock.mockImplementationOnce(() => false);
            rnaCacheHasMock.mockImplementationOnce(() => false);
            sendRequestMock.mockImplementationOnce(() => Promise.resolve(null))
            // @ts-ignore
            const actual = await apiAssoService.findFullScopeAssociation("ID");

            expect(actual).toBe(expected);
        })

        it("should return data with etablisements", async () => {

            sirenCacheHasMock.mockImplementationOnce(() => false);
            rnaCacheHasMock.mockImplementationOnce(() => false);
            sendRequestMock.mockImplementationOnce(() => Promise.resolve(fixtureAsso));
            const expected = fixtureEtablissements.length;
            // @ts-ignore
            const actual = (await apiAssoService.findFullScopeAssociation("ID"))?.etablissements.length;
            expect(actual).toEqual(expected)
        })

        it("should return data with documents", async () => {
            sirenCacheHasMock.mockImplementationOnce(() => false);
            rnaCacheHasMock.mockImplementationOnce(() => false);
            sendRequestMock.mockImplementationOnce(() => Promise.resolve(fixtureAsso));
            const expected = fixtureDocumentRna.length + fixtureDocumentDac.length;
            // @ts-ignore
            const actual = (await apiAssoService.findFullScopeAssociation("ID"))?.documents.length;
            expect(actual).toEqual(expected)
        })

        it("should save data in rna cache", async () => {
            sirenCacheHasMock.mockImplementationOnce(() => false);
            rnaCacheHasMock.mockImplementationOnce(() => false);
            sendRequestMock.mockImplementationOnce(() => Promise.resolve(fixtureAsso))
            // @ts-ignore
            const result = await apiAssoService.findFullScopeAssociation("ID");

            const expected = ["W00000000", result];
            expect(rnaCacheAddMock).toHaveBeenCalledWith(...expected);
        })

        it("should save data in siren cache", async () => {
            sirenCacheHasMock.mockImplementationOnce(() => false);
            rnaCacheHasMock.mockImplementationOnce(() => false);
            sendRequestMock.mockImplementationOnce(() => Promise.resolve(fixtureAsso))
            // @ts-ignore
            const result = await apiAssoService.findFullScopeAssociation("ID");

            const expected = ["509221941", result]
            expect(sirenCacheAddMock).toHaveBeenCalledWith(...expected);
        })
    })

    describe("Association Provider Part", () => {

        afterAll(() => {
            axiosMock.mockReset();
        })

        describe("getAssociationsBySiren", () => {

            beforeEach(() => {
                axiosMock.mockImplementationOnce(() => Promise.resolve({
                    status: 200,
                    data: fixtureAsso
                }))
            })

            it('should be return one association', async () => {
                const expected = 1;
                const actual = await apiAssoService.getAssociationsBySiren("509221941");

                expect(actual).toHaveLength(expected);
            })

            it('should be return association', async () => {
                const expected = [expect.objectContaining(fixtureAsso)];
                const actual = await apiAssoService.getAssociationsBySiren("509221941");

                expect(actual).toEqual(expected);
            })

            it('should be return null', async () => {
                const expected = null;
                //@ts-ignore
                jest.spyOn(apiAssoService, "findFullScopeAssociation").mockImplementationOnce(() => Promise.resolve(null))

                const actual = await apiAssoService.getAssociationsBySiren("00");
                expect(actual).toBe(expected);
            })
        });

        describe("getAssociationsBySiret", () => {

            beforeEach(() => {
                axiosMock.mockImplementationOnce(() => Promise.resolve({
                    status: 200,
                    data: fixtureAsso
                }))
            })

            it('should be return one association', async () => {
                const expected = 1;
                const actual = await apiAssoService.getAssociationsBySiret("50922194100000");

                expect(actual).toHaveLength(expected);
            })

            it('should be return association', async () => {
                const expected = [expect.objectContaining(fixtureAsso)];
                const actual = await apiAssoService.getAssociationsBySiret("50922194100000");

                expect(actual).toEqual(expected);
            })

            it('should be return null', async () => {
                const expected = null;
                //@ts-ignore
                jest.spyOn(apiAssoService, "findFullScopeAssociation").mockImplementationOnce(() => Promise.resolve(null))

                const actual = await apiAssoService.getAssociationsBySiret("00");
                expect(actual).toBe(expected);
            })
        });

        describe("getAssociationsByRna", () => {

            beforeEach(() => {
                axiosMock.mockImplementationOnce(() => Promise.resolve({
                    status: 200,
                    data: fixtureAsso
                }))
            })

            it('should be return one associations', async () => {
                const expected = 1;
                const actual = await apiAssoService.getAssociationsByRna("W0000000");

                expect(actual).toHaveLength(expected);
            })

            it('should be return association', async () => {
                const expected = [expect.objectContaining(fixtureAsso)];
                const actual = await apiAssoService.getAssociationsByRna("W0000000");

                expect(actual).toEqual(expected);
            })

            it('should be return null', async () => {
                const expected = null;
                //@ts-ignore
                jest.spyOn(apiAssoService, "findFullScopeAssociation").mockImplementationOnce(() => Promise.resolve(null))

                const actual = await apiAssoService.getAssociationsByRna("00");
                expect(actual).toBe(expected);
            })
        });
    })

    describe("Etablissement part", () => {
        afterAll(() => {
            axiosMock.mockReset();
        })
        beforeAll(() => {
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            apiAssoService.dataSirenCache.destroy();
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            apiAssoService.dataRnaCache.destroy();
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            apiAssoService.requestCache.destroy();

            axiosMock.mockImplementationOnce(() => Promise.resolve({
                status: 200,
                data: fixtureAsso
            }))
        })

        describe("getEtablissementsBySiret", () => {
            it('should be return one etablissement', async () => {
                const expected = 1;
                const actual = await apiAssoService.getEtablissementsBySiret("50922194100000");

                expect(actual).toHaveLength(expected);
            })

            it('should be return null', async () => {
                const expected = null;
                //@ts-ignore
                jest.spyOn(apiAssoService, "findFullScopeAssociation").mockImplementationOnce(() => Promise.resolve(null))

                const actual = await apiAssoService.getEtablissementsBySiret("00");
                expect(actual).toBe(expected);
            })
        })


        describe("getEtablissementsBySiren", () => {
            it('should be return two etablissements', async () => {
                const expected = 2;
                const actual = await apiAssoService.getEtablissementsBySiren("509221941");

                expect(actual).toHaveLength(expected);
            })

            it('should be return null', async () => {
                const expected = null;
                //@ts-ignore
                jest.spyOn(apiAssoService, "findFullScopeAssociation").mockImplementationOnce(() => Promise.resolve(null))

                const actual = await apiAssoService.getEtablissementsBySiren("00");
                expect(actual).toBe(expected);
            })
        })
    })

    describe("Documents part", () => {
        afterAll(() => {
            axiosMock.mockReset();
        })
        beforeAll(() => {
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            apiAssoService.dataSirenCache.destroy();
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            apiAssoService.dataRnaCache.destroy();
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            apiAssoService.requestCache.destroy();

            axiosMock.mockImplementationOnce(() => Promise.resolve({
                status: 200,
                data: fixtureAsso
            }))
        })

        describe("getDocumentsBySiret", () => {
            it('should be return one document', async () => {
                const expected = 1;
                const actual = await apiAssoService.getDocumentsBySiret("50922194100000");

                expect(actual).toHaveLength(expected);
            })

            it('should be return null', async () => {
                const expected = null;
                //@ts-ignore
                jest.spyOn(apiAssoService, "findFullScopeAssociation").mockImplementationOnce(() => Promise.resolve(null))

                const actual = await apiAssoService.getDocumentsBySiret("00");
                expect(actual).toBe(expected);
            })
        })


        describe("getDocumentsBySiren", () => {
            it('should be return 2 documents', async () => {
                const expected = 2;
                const actual = await apiAssoService.getDocumentsBySiren("509221941");

                expect(actual).toHaveLength(expected);
            })

            it('should be return null', async () => {
                const expected = null;
                //@ts-ignore
                jest.spyOn(apiAssoService, "findFullScopeAssociation").mockImplementationOnce(() => Promise.resolve(null))

                const actual = await apiAssoService.getDocumentsBySiren("00");
                expect(actual).toBe(expected);
            })
        })

        describe("getDocumentsByRna", () => {
            it('should be return 2 documents', async () => {
                const expected = 2;
                const actual = await apiAssoService.getDocumentsByRna("W00000000");

                expect(actual).toHaveLength(expected);
            })

            it('should be return null', async () => {
                const expected = null;
                //@ts-ignore
                jest.spyOn(apiAssoService, "findFullScopeAssociation").mockImplementationOnce(() => Promise.resolve(null))

                const actual = await apiAssoService.getDocumentsByRna("00");
                expect(actual).toBe(expected);
            })
        })
    })
});