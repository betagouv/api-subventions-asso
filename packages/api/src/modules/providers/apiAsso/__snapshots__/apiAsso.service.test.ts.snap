// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiAssoService findFullScopeAssociation should return data with etablisements and documents 1`] = `
Object {
  "associations": Array [
    Object {
      "activites": Object {},
      "coordonnees": Object {
        "adresse_siege": Object {
          "commune": "paris",
        },
      },
      "date_modification": Array [
        Object {
          "value": 2022-01-01T00:00:00.000Z,
        },
      ],
      "denomination": Array [
        Object {
          "provider": "TEST",
          "value": "TEST",
        },
      ],
      "document_dac": Array [
        Object {
          "meta": Object {
            "etat": "courant",
            "id_siret": "50922194100000",
            "type": "BPA",
          },
          "nom": "nom fake",
          "time_depot": "2021-06-18 12:02:53",
          "url": "/fake/url",
        },
      ],
      "document_rna": Array [
        Object {
          "annee": 2021,
          "id": "ididididid",
          "sous-type": "PV",
          "time": 1622557171,
          "type": "PV",
          "url": "/fake/url",
        },
      ],
      "etablissement": Array [
        Object {
          "adresse": Object {},
          "id_siret": "50922194100000",
        },
        Object {
          "adresse": Object {},
          "id_siret": "50922194100001",
        },
      ],
      "identite": Object {
        "date_modif_rna": "01-01-2022",
        "date_modif_siren": "01-01-2022",
        "id_forme_juridique": "9220",
        "id_rna": "W00000000",
        "id_siren": "509221941",
        "id_siret_siege": "5092219410000",
        "nom": "TEST",
        "nom_sirene": "TEST SIREN 2",
        "regime": "Autre",
      },
      "rib": Array [
        Object {
          "bic": "BIC_TEST",
          "iban": "TEST",
          "id_siret": "50922194100000",
        },
        Object {
          "bic": "BIC_TEST 2",
          "iban": "TEST 2",
          "id_siret": "50922194100000",
          "url": "fake/path/to/rib",
        },
      ],
    },
  ],
  "documents": Array [
    Object {
      "__meta__": Object {},
      "nom": Object {
        "last_update": 2021-01-19T18:42:37.171Z,
        "provider": "BASE RNA <Via API ASSO>",
        "type": "string",
        "value": "ididididid",
      },
      "type": Object {
        "last_update": 2021-01-19T18:42:37.171Z,
        "provider": "BASE RNA <Via API ASSO>",
        "type": "string",
        "value": "PV",
      },
      "url": Object {
        "last_update": 2021-01-19T18:42:37.171Z,
        "provider": "BASE RNA <Via API ASSO>",
        "type": "string",
        "value": "/fake/url",
      },
    },
    Object {
      "__meta__": Object {
        "siret": "50922194100000",
      },
      "nom": Object {
        "last_update": 2021-06-18T12:02:00.000Z,
        "provider": "Document déposer dans le comtpe asso <Via API ASSO>",
        "type": "string",
        "value": "nom fake",
      },
      "type": Object {
        "last_update": 2021-06-18T12:02:00.000Z,
        "provider": "Document déposer dans le comtpe asso <Via API ASSO>",
        "type": "string",
        "value": "BPA",
      },
      "url": Object {
        "last_update": 2021-06-18T12:02:00.000Z,
        "provider": "Document déposer dans le comtpe asso <Via API ASSO>",
        "type": "string",
        "value": "/fake/url",
      },
    },
  ],
  "etablissements": Array [
    Object {
      "adresse": Object {},
      "id_siret": "50922194100000",
      "siret": Array [
        Object {
          "value": "50922194100000",
        },
      ],
    },
    Object {
      "adresse": Object {},
      "id_siret": "50922194100001",
      "siret": Array [
        Object {
          "value": "50922194100001",
        },
      ],
    },
  ],
}
`;
